src/transformers/pipelines/zero_shot_object_detection.py

```diff
@@ -142,49 +140,54 @@ def _sanitize_parameters(self, **kwargs):
-            text_queries = [text_queries]
-
-        target_sizes = [torch.IntTensor([[img.height, img.width]]) for img in images]
-        target_sizes = torch.cat(target_sizes)
-        inputs = self._processor(text=inputs["text_queries"], images=images, return_tensors="pt")
-        return {"target_sizes": target_sizes, "text_queries": text_queries, **inputs}
+        image = load_image(inputs["image"])
+        candidate_labels = inputs["candidate_labels"]
+        if isinstance(candidate_labels, str):
+            candidate_labels = candidate_labels.split(",")
+
+        target_size = torch.tensor([[image.height, image.width]], dtype=torch.int32)
+        for i, candidate_label in enumerate(candidate_labels):
+            text_inputs = self.tokenizer(candidate_label, return_tensors=self.framework)
+            image_features = self.feature_extractor(image, return_tensors=self.framework)
+            yield {
+                "is_last": i == len(candidate_labels) - 1,
+                "target_size": target_size,
+                "candidate_label": candidate_label,
+                **text_inputs,
+                **image_features,
+            }
 
     def _forward(self, model_inputs):
-        target_sizes = model_inputs.pop("target_sizes")
-        text_queries = model_inputs.pop("text_queries")
+        target_size = model_inputs.pop("target_size")
+        candidate_label = model_inputs.pop("candidate_label")
+        is_last = model_inputs.pop("is_last")
+
         outputs = self.model(**model_inputs)
 
-        model_outputs = outputs.__class__({"target_sizes": target_sizes, "text_queries": text_queries, **outputs})
+        model_outputs = {"target_size": target_size, "candidate_label": candidate_label, "is_last": is_last, **outputs}
         return model_outputs
 
     def postprocess(self, model_outputs, threshold=0.1, top_k=None):
-        texts = model_outputs["text_queries"]
-
-        outputs = self.feature_extractor.post_process(
-            outputs=model_outputs, target_sizes=model_outputs["target_sizes"]
-        )
 
         results = []
-        for i in range(len(outputs)):
-            keep = outputs[i]["scores"] >= threshold
-            labels = outputs[i]["labels"][keep].tolist()
-            scores = outputs[i]["scores"][keep].tolist()
-            boxes = [self._get_bounding_box(box) for box in outputs[i]["boxes"][keep]]
-
-            result = [
-                {"score": score, "label": texts[i][label], "box": box}
-                for score, label, box in zip(scores, labels, boxes)
-            ]
-
-            result = sorted(result, key=lambda x: x["score"], reverse=True)
-            if top_k:
-                result = result[:top_k]
-            results.append(result)
+        for model_output in model_outputs:
+            label = model_output["candidate_label"]
+            model_output = BaseModelOutput(model_output)
+            outputs = self.feature_extractor.post_process(
```

Comment:
@Narsil This is where I assume the non-max suppression would be done. We can check once though

Reply:
It was in the old code though: https://github.com/huggingface/transformers/pull/20280/files#diff-3ffc4527266bf533daeecb267c6e555a86fb8acd36c4207aecffb08c76b1f656L168

Reply:
Happy to switch though, this will need rework since I guess we need to stack all the outputs together before calliung this.

Is `post_process_image_guided_detection` defined for any other feature_extractor and/or is it easily implementable by any model that would loook like Owl ? (Just want to make sure this pipeline is not overfitting a single model)

Reply:
The old code only had the probability threshold but no NMS, could you add it back @Narsil? 

`post_process_image_guided_detection` is only implemented for OWL-ViT but it is applicable to all object detection post-process methods. I can open a separate PR to add NMS to the post-process methods of all object detection models, including OWL-ViT's zero-shot post-process method.

Reply:
If there's nothing specific about owl, We could also have the function directly within the pipeline. 

Or a util function that both the `feature_extractor` and the `pipeline could` use maybe ? :D.

Reply:
It depends, supporting NMS in all post-processing methods would be really handy and I'll add it to all relevant models' feature extractors anyway. Or as you suggested, I can create a util function and edit the post-processing methods.

Reply:
Whichever your prefer. 

I like indedendant functions usually because there is less state to them (no `self.xxx` ) so it's easier to reason about what they do. Ofc this only works if the amount of arguments is limited.

Reply:
Makes sense, I will go for the common util function for NMS + improved post-processing methods with `nms_threshold=0` for backward compatibility then.

Reply:
I will merge this and we can fix NMS in another PR, ok ?

Reply:
Sounds good!